File: src/com/android/settings/network/telephony/gsm/OpenNetworkSelectPagePreferenceController.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.settings.network.telephony.gsm;

import static androidx.lifecycle.Lifecycle.Event.ON_START;
import static androidx.lifecycle.Lifecycle.Event.ON_STOP;

import android.content.Context;
import android.content.Intent;
import android.provider.Settings;
import android.telephony.ServiceState;
import android.telephony.SubscriptionManager;
import android.telephony.TelephonyManager;

import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.preference.Preference;
import androidx.preference.PreferenceScreen;

import com.android.settings.R;
import com.android.settings.network.AllowedNetworkTypesListener;
import com.android.settings.network.SubscriptionsChangeListener;
import com.android.settings.network.telephony.DomesticRoamUtils;
import com.android.settings.network.telephony.Enhanced4gBasePreferenceController;
import com.android.settings.network.telephony.MobileNetworkUtils;
import com.android.settings.network.telephony.TelephonyBasePreferenceController;

/**
 * Preference controller for "Open network select"
 */
public class OpenNetworkSelectPagePreferenceController extends
        TelephonyBasePreferenceController implements
        AutoSelectPreferenceController.OnNetworkSelectModeListener,
        Enhanced4gBasePreferenceController.On4gLteUpdateListener,
        LifecycleObserver,
        SubscriptionsChangeListener.SubscriptionsChangeListenerClient {

    private TelephonyManager mTelephonyManager;
    private Preference mPreference;
    private PreferenceScreen mPreferenceScreen;
    private AllowedNetworkTypesListener mAllowedNetworkTypesListener;
    private int mCacheOfModeStatus;
    private SubscriptionsChangeListener mSubscriptionsListener;

    public OpenNetworkSelectPagePreferenceController(Context context, String key) {
        super(context, key);
        mTelephonyManager = context.getSystemService(TelephonyManager.class);
        mSubId = SubscriptionManager.INVALID_SUBSCRIPTION_ID;
        mCacheOfModeStatus = TelephonyManager.NETWORK_SELECTION_MODE_UNKNOWN;
        mAllowedNetworkTypesListener = new AllowedNetworkTypesListener(
                context.getMainExecutor());
        mAllowedNetworkTypesListener.setAllowedNetworkTypesListener(
                () -> updatePreference());
        mSubscriptionsListener = new SubscriptionsChangeListener(context, this);

    }

    @Override
    public void on4gLteUpdated() {
        updateState(mPreference);
    }

    private void updatePreference() {
        if (mPreferenceScreen != null) {
            displayPreference(mPreferenceScreen);
        }
        if (mPreference != null) {
            updateState(mPreference);
        }
    }

    @Override
    public int getAvailabilityStatus(int subId) {
        return MobileNetworkUtils.shouldDisplayNetworkSelectOptions(mContext, subId)
                ? AVAILABLE
                : CONDITIONALLY_UNAVAILABLE;
    }

    @OnLifecycleEvent(ON_START)
    public void onStart() {
        mAllowedNetworkTypesListener.register(mContext, mSubId);
        mSubscriptionsListener.start();
    }

    @OnLifecycleEvent(ON_STOP)
    public void onStop() {
        mAllowedNetworkTypesListener.unregister(mContext, mSubId);
        mSubscriptionsListener.stop();
    }

    @Override
    public void displayPreference(PreferenceScreen screen) {
        super.displayPreference(screen);
        mPreferenceScreen = screen;
        mPreference = screen.findPreference(getPreferenceKey());
    }

    @Override
    public void updateState(Preference preference) {
        super.updateState(preference);
        final int phoneType = mTelephonyManager.getPhoneType();
        if (phoneType == TelephonyManager.PHONE_TYPE_CDMA) {
             preference.setEnabled(false);
        } else {
        preference.setEnabled(mCacheOfModeStatus
                != TelephonyManager.NETWORK_SELECTION_MODE_AUTO);
        }

        Intent intent = new Intent();
        intent.setClassName("com.android.settings",
                "com.android.settings.Settings$NetworkSelectActivity");
        intent.putExtra(Settings.EXTRA_SUB_ID, mSubId);
        preference.setIntent(intent);
    }

    @Override
    public CharSequence getSummary() {
        final ServiceState ss = mTelephonyManager.getServiceState();
        if (ss != null && ss.getState() == ServiceState.STATE_IN_SERVICE) {
            if (DomesticRoamUtils.isFeatureEnabled(mContext)) {
                String registeredOperatorName = DomesticRoamUtils.getRegisteredOperatorName(
                        mContext, mSubId);
                if (DomesticRoamUtils.EMPTY_OPERATOR_NAME != registeredOperatorName) {
                    return registeredOperatorName;
                }
            }
            return MobileNetworkUtils.getCurrentCarrierNameForDisplay(mContext, mSubId);
        } else {
            return mContext.getString(R.string.network_disconnected);
        }
    }

    /**
     * Initialization based on given subscription id.
     **/
    public OpenNetworkSelectPagePreferenceController init(int subId) {
        mSubId = subId;
        mTelephonyManager = mContext.getSystemService(TelephonyManager.class)
                .createForSubscriptionId(mSubId);
        return this;
    }

    @Override
    public void onNetworkSelectModeUpdated(int mode) {
        mCacheOfModeStatus = mode;
        if (mPreference != null) {
            updateState(mPreference);
        }
    }

    @Override
    public void onAirplaneModeChanged(boolean airplaneModeEnabled) {
    }

    @Override
    public void onSubscriptionsChanged() {
        updateState(mPreference);
    }
}

File: src/com/android/settings/network/telephony/gsm/OpenNetworkSelectPagePreferenceController.kt
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// QTI_BEGIN: 2024-03-14: Telephony: CAG and SNPN feature
/*
 * Changes from Qualcomm Innovation Center are provided under the following license:
 *
 * Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause-Clear
 */

// QTI_END: 2024-03-14: Telephony: CAG and SNPN feature
package com.android.settings.network.telephony.gsm

import android.content.Context
import android.content.Intent
import android.provider.Settings
import android.telephony.ServiceState
import android.telephony.TelephonyManager
// QTI_BEGIN: 2024-03-17: Telephony: UI requirement in CU domestic roaming
import android.util.Log
// QTI_END: 2024-03-17: Telephony: UI requirement in CU domestic roaming
import androidx.lifecycle.LifecycleOwner
import androidx.preference.Preference
import androidx.preference.PreferenceScreen
import com.android.settings.R
import com.android.settings.Settings.NetworkSelectActivity
// QTI_BEGIN: 2024-03-17: Telephony: UI requirement in CU domestic roaming
import com.android.settings.network.telephony.DomesticRoamUtils
// QTI_END: 2024-03-17: Telephony: UI requirement in CU domestic roaming
import com.android.settings.network.telephony.MobileNetworkUtils
import com.android.settings.network.telephony.TelephonyBasePreferenceController
import com.android.settings.network.telephony.allowedNetworkTypesFlow
import com.android.settings.network.telephony.serviceStateFlow
import com.android.settingslib.spa.framework.util.collectLatestWithLifecycle
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.withContext

// QTI_BEGIN: 2024-03-14: Telephony: CAG and SNPN feature
import com.qti.extphone.ExtTelephonyManager

// QTI_END: 2024-03-14: Telephony: CAG and SNPN feature
/** Preference controller for "Open network select" */
class OpenNetworkSelectPagePreferenceController
@JvmOverloads
constructor(
    context: Context,
    key: String,
    private val allowedNetworkTypesFlowFactory: (subId: Int) -> Flow<Long> =
        context::allowedNetworkTypesFlow,
    private val serviceStateFlowFactory: (subId: Int) -> Flow<ServiceState> =
        context::serviceStateFlow,
) :
    TelephonyBasePreferenceController(context, key),
    AutoSelectPreferenceController.OnNetworkSelectModeListener {

    private var preference: Preference? = null
    private var networkSelectionMode: Int = TelephonyManager.NETWORK_SELECTION_MODE_AUTO
// QTI_BEGIN: 2024-03-14: Telephony: CAG and SNPN feature
    private val telephonyManager = context.getSystemService(TelephonyManager::class.java)!!
// QTI_END: 2024-03-14: Telephony: CAG and SNPN feature

    /** Initialization based on given subscription id. */
    fun init(subId: Int): OpenNetworkSelectPagePreferenceController {
        mSubId = subId
// QTI_BEGIN: 2024-03-14: Telephony: CAG and SNPN feature
        telephonyManager.createForSubscriptionId(mSubId)
// QTI_END: 2024-03-14: Telephony: CAG and SNPN feature
        return this
    }

    override fun updateState(preference: Preference?) {
        super.updateState(preference)
        preference?.isEnabled =
            !mIsAirplaneModeOn &&
                networkSelectionMode != TelephonyManager.NETWORK_SELECTION_MODE_AUTO
    }

    override fun getAvailabilityStatus(subId: Int) =
        if (MobileNetworkUtils.shouldDisplayNetworkSelectOptions(mContext, subId)) AVAILABLE
        else CONDITIONALLY_UNAVAILABLE

    override fun displayPreference(screen: PreferenceScreen) {
        super.displayPreference(screen)
        preference = screen.findPreference(preferenceKey)
        preference?.intent =
            Intent().apply {
                setClass(mContext, NetworkSelectActivity::class.java)
                putExtra(Settings.EXTRA_SUB_ID, mSubId)
            }
    }

    override fun onViewCreated(viewLifecycleOwner: LifecycleOwner) {
        allowedNetworkTypesFlowFactory(mSubId).collectLatestWithLifecycle(viewLifecycleOwner) {
            preference?.isVisible =
                withContext(Dispatchers.Default) {
                    MobileNetworkUtils.shouldDisplayNetworkSelectOptions(mContext, mSubId)
                }
        }

        serviceStateFlowFactory(mSubId).collectLatestWithLifecycle(viewLifecycleOwner) {
            serviceState ->
            preference?.summary =
                if (serviceState.state == ServiceState.STATE_IN_SERVICE ||
// QTI_BEGIN: 2024-03-14: Telephony: CAG and SNPN feature
                        isSnpnInService(serviceState)) {
// QTI_END: 2024-03-14: Telephony: CAG and SNPN feature
                    withContext(Dispatchers.Default) {
// QTI_BEGIN: 2024-03-17: Telephony: UI requirement in CU domestic roaming
                        if (DomesticRoamUtils.isFeatureEnabled(mContext)) {
                            val registeredOperatorName : String = DomesticRoamUtils
                                    .getRegisteredOperatorName(mContext, mSubId)
                            if (DomesticRoamUtils.EMPTY_OPERATOR_NAME != registeredOperatorName) {
                                registeredOperatorName
// QTI_END: 2024-03-17: Telephony: UI requirement in CU domestic roaming
// QTI_BEGIN: 2024-03-19: Telephony: UI requirement in CU domestic roaming
                            } else {
                                MobileNetworkUtils.getCurrentCarrierNameForDisplay(mContext, mSubId)
// QTI_END: 2024-03-19: Telephony: UI requirement in CU domestic roaming
// QTI_BEGIN: 2024-03-17: Telephony: UI requirement in CU domestic roaming
                            }
// QTI_END: 2024-03-17: Telephony: UI requirement in CU domestic roaming
// QTI_BEGIN: 2024-03-19: Telephony: UI requirement in CU domestic roaming
                        } else {
                            MobileNetworkUtils.getCurrentCarrierNameForDisplay(mContext, mSubId)
// QTI_END: 2024-03-19: Telephony: UI requirement in CU domestic roaming
// QTI_BEGIN: 2024-03-17: Telephony: UI requirement in CU domestic roaming
                        }
// QTI_END: 2024-03-17: Telephony: UI requirement in CU domestic roaming
                    }
                } else {
                    mContext.getString(R.string.network_disconnected)
                }
        }
    }

// QTI_BEGIN: 2024-03-14: Telephony: CAG and SNPN feature
    private fun isSnpnInService(ss: ServiceState): Boolean {
        return ((MobileNetworkUtils.getAccessMode(mContext, telephonyManager.getSlotIndex())
                == ExtTelephonyManager.ACCESS_MODE_SNPN)
                && (ss.getDataRegState() == ServiceState.STATE_IN_SERVICE))
    }

// QTI_END: 2024-03-14: Telephony: CAG and SNPN feature
    override fun onNetworkSelectModeUpdated(mode: Int) {
        this.networkSelectionMode = mode
        updateState(preference)
    }
}
